rfcs/framework-0035-error-handling-and-response.md

```diff
@@ -0,0 +1,323 @@
+
+**Flask's error handling/rendering**
+
+Flask and Werkzeug's request dispatching system will catch all errors happening in the three layers described. The error handling is pretty sophisticated, and is explained below.
+
+First, Flask will look for a user-provided error handler. If found, the handler is responsible for making a response. If not found, but the exception is an HTTPException (or subclass thereof), the HTTPException will simply be returned, because an HTTPException can self-render. If it's not an HTTPException, the error is simply reraised and results in an Internal Server Error. The following pseudo-code is an example of the handling:
+
+```python
+def handle_exception(e):
+    # If it's an HTTPException, use different method:
+    if isinstance(e, HTTPException):
+        return handle_http_exception(e)
+
+    handler = find_error_handler(e)
+    if handler is None:
+        # No handler was found, and it's not and HTTPException.
+        raise # <----- reraise
+    # A handler was found, use it to return response
+    return handler(e)
+
+def handle_http_exception(e):
+    handler = find_error_handler(e)
+    if handler is None:
+        # No handler was found, but it is an HTTPException.
+        return e # <----- return
+    # A handler was found, use it to return response
+    return handler(e)
+```
+
+The key in understanding the difference between an normal exception and an HTTPException is that an HTTPException is considered a valid response by Flask's ``make_reponse``. This allows it to be returned by the ``handle_http_exception()`` function.
+
+**Recap**
+
+Source of errors (exceptions raised) is three layers (Flask, View, Service). For handling these errors (i.e. making an HTTP error response) we can use Flask and Werkzeug's error handling system. The view layer is not well-suited for this.
+
+Above basically establish that either
+
+1. an *exception* can render itself into an HTTP error response, or
+2. an *error handler* can render an exception into a HTTP error response.
+
+### Error handling in Flask REST API applications
+
+Following is a proposal for how to implement the error handling in REST API application
+
+**JSON-formatted HTTP Exception**
+
+First we propose to create a new subclass of ``HTTPException``, that will render the response body as JSON. In simplified pseudo-Python  this would be something like:
+
+
+```python
+class HTTPJSONException(HTTPException):
+    def get_body(self, environ=None):
+        body = dict(
+            status=self.code,
+            message=self.get_description(environ=environ),
+        )
+        return json.dumps(body)):
+```
+
+**Flask application**
+
+Next, we install an error handler on the Flask application for ``HTTPExceptions``:
+
+```python
+from Flask import Flask
+
+app = Flask(__name__)
+
+@app.errorhandler(HTTPException)
```

Comment:
I think that's the core of my disagreement: in invenio-records-resources this can be done at the app level but in flask-resources this should be done on a blueprint **only**.

Reply:
The RFC is about error handling principles in an Invenio-based application, and it is not about if Flask-Resources or Invenio-Records-Resources should install application-level error handlers. The main purpose is to document the thoughts behind the *current* approach.

Already today, Invenio-REST is installing the application-level error handlers, and Invenio-Records-REST is installing blueprint level error handlers more or less according to what is described here. There are few details in the current implementation that could be simplified and improved, but overall the approach is the documented one. So you can say this RFC is simply confirming the current approach. 

I think it's probably more a misunderstanding of scope that needs to better documented in the RFC.

Reply:
Probably the scope is throwing me off here :sweat_smile: . Is this how we want error handling to be managed in the future (in flask-resources and invenio-records-resources) or is it how it *is* working in production invenio (invenio-rest)? Maybe the RFC cuts things differently and that is not even a concern... :thinking: ?